src/Shared/Components/ComponentDescriptorSerializer.Deserialize.cs

```diff
@@ -0,0 +1,215 @@
+    // This file holds the definition used by Blazor to deserialize descriptors.
+    internal partial class ComponentDescriptorSerializer
+    {
+        public ComponentDescriptorSerializer(
+            IDataProtectionProvider dataProtectionProvider,
+            ILogger<ComponentDescriptorSerializer> logger,
+            RootComponentTypeCache rootComponentTypeCache)
+        {
+            DataProtector = dataProtectionProvider.CreateProtector(DataProtectionProviderPurpose);
+            Logger = logger;
+            RootComponentTypeCache = rootComponentTypeCache;
+        }
+
+        public ILogger<ComponentDescriptorSerializer> Logger { get; }
+
+        public RootComponentTypeCache RootComponentTypeCache { get; }
+
+        public bool TryDeserializeComponentDescriptorCollection(string serializedComponentRecords, out List<ComponentDescriptor> descriptors)
+        {
+            var markers = JsonSerializer.Deserialize<IEnumerable<ComponentMarker>>(serializedComponentRecords, _jsonSerializationOptions);
+            descriptors = new List<ComponentDescriptor>();
+            int lastSequence = -1;
+
+            var previousInstance = new ComponentDescriptorInstance();
+            foreach (var marker in markers)
+            {
+                if (marker.Type != ComponentMarker.ServerMarkerType)
+                {
+                    Log.InvalidMarkerType(Logger, marker.Type);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                if (marker.Sequence == null)
+                {
+                    Log.MissingMarkerSequence(Logger);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                if (marker.Descriptor == null)
+                {
+                    Log.MissingMarkerDescriptor(Logger);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                var (descriptor, instance) = DeserializeComponentDescriptor(marker);
+                if (descriptor == null)
+                {
+                    // We failed to deserialize the component descriptor for some reason.
+                    descriptors.Clear();
+                    return false;
+                }
+
+                // We force our client to send the descriptors in order so that we do minimal work.
+                if (lastSequence != -1 && lastSequence != instance.Sequence - 1)
+                {
+                    Log.OutOfSequenceDescriptor(Logger, lastSequence, instance.Sequence);
+                    descriptors.Clear();
+                    return false;
+                }
+
+                if (lastSequence != -1 &&
+                    !CryptographicOperations.FixedTimeEquals(
+                        MemoryMarshal.AsBytes(previousInstance.InvocationId.AsSpan()),
+                        MemoryMarshal.AsBytes(instance.InvocationId.AsSpan())))
+                {
```

Comment:
I believe we can actually perform a regular comparison here. This is all protected data not susceptible to user input, so at best, sending valid descriptors an attacker could identify that the payload in their descriptor is a prefix of the payload in the other descriptor, but given that they don't have control over the input (they don't know their InvocationId nor the one in the other descriptor and can't generate payloads) its in practical terms impossible for them to exploit that for anything

Reply:
We agreed we don't need to do it this way